[Step3]구간 추가 기능 PR 요청 드립니다.

re-mora opened the pull request:
안녕하세요 :) 교육생 최용석입니다 😄 

3단계 구간 추가 기능 미션 구현이 완료되어, 코드 리뷰 요청드립니다.

이번 단계는 지하철 노선과 구간이라는 개념을 이해하고 구현하는 과정에 많은 어려움이 있었지만
최대한 `구간목록` 이라는 일급컬렉션에 최대한 비지니스 로직을 표현해보고자 했고,
JPA의 연관관계 매핑, 영속성 전이, Lazy Loading + Batch Size를 이용한 발생 쿼리 개선 등에 주안하여 구현해 보았습니다!

또한 구간 추가라는 기능이 구현되기 위한 여러 케이스에 대한 인수테스트를 작성하였고, 
예외 발생 케이스에 대한 인수테스트 역시 같이 작성해 보았습니다.

- [구간 추가에 대한 여러 인수테스트 작성](https://github.com/next-step/atdd-subway-admin/pull/823/commits/64ffaba1354074faedb12c994ebfcb880ff82a5e)

이번 미션 역시 미흡한 부분이 있다면 피드백 부탁드립니다 :)

re-mora committed:
docs: 요구사항 분석 및 구현 기능 목록 작성

re-mora committed:
refactor: 각 인수테스트 API 응답의 id에 해당하는 자료형을 Long으로 변경

re-mora committed:
test: 실패하는 지하철 노선 구간 추가 인수테스트 작성

re-mora committed:
feat: 구간에 대한 Entity 설계

- 노선에 대한 일대다 연관관계 설정
- 상행역, 하행역에 대한 일대다 연관관계 설정

re-mora committed:
feat: 구간 목록에 대한 일급 컬렉션 정의

- @Embeddable을 이용한 컬렉션 타입의 Entity 정의

re-mora committed:
feat: 노선과 구간의 연관관계 설정

- @Embedded를 이용한 컬렉션 타입의 필드 정의

re-mora committed:
chore(equalsAndHashCode): Station 객체의 equals(), hashCode() 추가

re-mora committed:
docs: 구현 기능 목록 수정

re-mora committed:
docs: 구현 기능 목록 수정

re-mora committed:
feat: Domain 레벨에서의 상행/하행 종점역 구간 추가 기능 및 TC 작성

구현 기능
- 상행/하행역 목록 조회
- 상행/하행 종점역 조회
  - 상행종점역 : 하행역 목록과 상행역 목록의 차집합
  - 하행종점역 : 상행역 목록과 하행역 목록의 차집합
- 상행/하행 종점역 구간 조회
- 노선에 포함된 정렬된 역 목록 조회
- `신규 상행/하행 종점역 구간 추가`

문서 작성
- 추가 구현 기능 및 예외 추가
- 기능 구현에 따른 TODO List 갱신

임시로 커밋했던 작성되지 않은 불필요 파일 제거

re-mora committed:
feat: Domain 레벨에서의 중간 구간 추가 기능 및 TC 작성

구현 기능
- 상행역이 동일한 구간 탐색 및 하행역 변경
- 하행역이 동알한 구간 탐색 및 상행역 변경

문서 작성
- 구간 추가 기능의 여러 케이스에 대한 구현 예시 작성
- 기능 구현에 따른 TODO List 갱신

re-mora committed:
feat: 구간 추가 시 발생할 수 있는 예외 처리

- 동일 구간 추가 요청 시 예외 처리 및 검증 TC 작성
- 구간 추가 시, 인접한 역이 없는 경우 예외 처리 및 검증 TC 작성
- 구간 거리가 유효하지 못한 경우 예외 및 검증 TC 작성
  - 추가 구간의 거리가 전체 구간의 거리보다 같거나 큰 경우
  - 기존 구간 사이에 구간을 추가 할 때, 기존 구간의 길이보다 같거나 큰 경우
  - 원시 타입 포장 객체를 이용하여 구간 생성 시점에 유효하지 못한 거리에 대한 유효성 처리
- 전체 구간의 총 길이 반환 로직 구현

테스트 가독성을 위한 표현 개선

문서 작성
- 기능 구현에 따른 TODO List 갱신

re-mora committed:
refactor: 구간 생성 시 발생할 수 있는 예외 처리

예외 처리 및 검증 TC 작성
- 상행역과 하행역이 동일한 구간 생성 시 예외
- 상행역 혹은 하행역 정보가 없는 구간 생성 시 예외
- 상행역과 하행역이 동일한 구간 생성 시 예외
- 유효하지 않는 길이를 가지는 구간 생성 시 에외

정적 팩토리 메서드를 통해 유효성 검증이 완료된 객체 생성 유도

re-mora committed:
refactor: 초기 노선 설계 시 설정했던 불필요 연관관계 삭제 및 누락된 기능에 대한 TC 작성

노선의 단일 구간 설계로 인해 `일대일`로 설정된 연관관계 제거
- 연관관계 제거에 따른 코드 수정

구간 추가 시, 변경감지 및 영속성 전이에 따른 동작 검증 TC 작성

노선 목록 조회 시, Lazy Loading + Batch Size 설정을 이용한 조회 쿼리 개선

re-mora committed:
refactor: 구간 추가 API 구현 및 구간 추가에 대한 인수테스트 성공 검증

re-mora committed:
chore(re-package): 명확하지 않은 package명 수정

- AS-IS : station
- TO-BE : acceptance

re-mora committed:
test: 구간 추가에 대한 여러 인수테스트 작성

구간 추가에 대한 인수테스트
- 지하철 노선에 신규 상행종점역 구간을 추가한다.
- 지하철 노선에 신규 하행종점역 구간을 추가한다.
- 지하철 노선에 상행역이 동일한 구간을 추가한다.
- 지하철 노선에 하행역이 동일한 구간을 추가한다.

구간 추가 시 예외가 발생하는 인수테스트
- 동일한 구간을 추가하는 경우 예외가 발생한다.
- 접점이 없는 역이 포함된 구간을 추가하는 경우 예외가 발생한다.
- 구간의 길이가 유효하지 않은 경우 예외가 발생한다.

parkeeseul commented on src/main/java/nextstep/subway/ui/LineController.java:
```diff
@@ -55,4 +56,13 @@ public ResponseEntity deleteLineById(@PathVariable(value = "id") Long id) {
+        LineResponse lineResponse = lineService.addSection(id, createSectionRequest);
+        return ResponseEntity.created(URI.create("/lines/" + lineResponse.getId())).body(lineResponse);
+    }
 }
```
요구사항에 있는 API 명세를 확인해 주세요! 😄 

`POST /lines/1/sections`
![스크린샷 2022-06-22 오후 6 32 39](https://user-images.githubusercontent.com/48339310/174996016-b92ccf70-ad50-4cac-a3f6-cd5f41db71db.png)

re-mora replied:
요구사항을 꼼꼼하게 확인 하지 못했네요..

구간 추가 API의 요청 경로를 아래와 같이 수정하였습니다.
- AS-IS : /lines/{id}
- TO-BE : /lines/{id}/sections

parkeeseul commented on src/test/java/nextstep/subway/acceptance/SectionAcceptanceTest.java:
```diff
@@ -0,0 +1,201 @@
+     * Then 접점이 없는 역이 포함된 구간 추가 시 오류가 발생한다.
+     */
+    @Test
+    @DisplayName("접점이 없는 역이 포함된 구간을 추가하는 경우 예외가 발생한다.")
```
요구사항이 그대로 테스트 코드가 검증해야 하는 목적이 되는 경우가 많다고 생각하는데요,
때문에 개인적으로는 요구사항을 그대로 테스트 설명을 작성해 주는 것이 
요구사항과 테스트 코드를 함께 확인하기 좋다고 생각합니다.
어떻게 생각하시나요? 😄 
`상행역과 하행역 둘 중 하나도 포함되어있지 않으면 추가할 수 없음`

re-mora replied:
테스트 내용을 자세하게 표현하려다 보니 오히려 읽기 어려운 문장이 된 것 같습니다. 😭 
말씀 하신것처럼 요구사항을 기반으로 해당 테스트에 대한 설명을 아래와 같이 수정해 보았습니다

`추가 구간의 상행/하행역이 모두 노선에 존재하지 않는 경우 추가할 수 없다.`

parkeeseul commented on src/test/java/nextstep/subway/acceptance/SectionAcceptanceTest.java:
```diff
@@ -0,0 +1,201 @@
+     */
+    @ParameterizedTest
+    @ValueSource(ints = {Integer.MIN_VALUE, 0, 50, 51, 100, 101})
+    @DisplayName("구간의 길이가 유효하지 않은 경우 예외가 발생한다.")
```
구간의 길이가 어떤 경우 유효하지 않은 경우일까요? 🤔 
조금더 구체적으로 작성해 주면 좋을 것 같아요 😄

re-mora replied:
구간의 길이가 유효하지 않은 경우를 조금 모호하게 표현한 것 같습니다.
말씀하신 것처럼 구간의 길이가 유효하지 않은 케이스를 아래와 같이 최대한 표현해 보았습니다. 

하나의 테스트에서 여러 경우의 수가 수행될 때, 어떤 테스트인지 표현하는게 어려운 부분인 것 같습니다.. 😭  

```java
    /**
     * Given `논현-(50)-신논현-(50)-정자` 구간이 존재하는 노선을 생성한다.
     * When 구간의 길이가 유효하지 않은 구간을 추가한다.
     *      When - 추가 구간의 길이가 0 혹은 음수인 경우
     *      When - 추가 구간의 길이가 연결 구간의 길이보다 크거나 같은 경우
     *      When - 추가 구간의 길이가 전체 구간의 길이보다 크거나 같은 경우
     * Then 구간의 길이가 유효하지 않는 경우 오류가 발생한다.
     */
    @ParameterizedTest
    @MethodSource
    @DisplayName("추가 구간의 길이가 `0, 음수, 연결 구간의 길이보다 크거나 같은 경우, 전체 구간의 길이보다 크거나 같은 경우`는 추가할 수 없다.")
    public void throwException_WhenAddInvalidSectionDistance(int 유효하지_않은_구간_길이, String givenDescription) {
        // Given
        Long 신논현역 = getIdAsLong(지하철역_생성_요청("신논현역"));
        Long 강남역 = getIdAsLong(지하철역_생성_요청("강남역"));
        지하철_구간_생성_요청(신분당선, new CreateSectionRequest(논현역, 신논현역, 50));

        // When
        Response 지하철_구간_생성_응답 = 지하철_구간_생성_요청(신분당선, new CreateSectionRequest(강남역, 정자역, 유효하지_않은_구간_길이));

        // Then
        assertInternalServerError(지하철_구간_생성_응답);
    }

    private static Stream throwException_WhenAddInvalidSectionDistance() {
        return Stream.of(
            Arguments.of(Integer.MIN_VALUE, "추가 구간의 길이가 음수인 경우"),
            Arguments.of(0, "추가 구간의 길이가 0인 경우"),
            Arguments.of(50, "추가 구간의 길이가 연결 구간의 길이와 같은 경우"),
            Arguments.of(51, "추가 구간의 길이가 연결 구간의 길이보다 큰 경우"),
            Arguments.of(100, "추가 구간의 길이가 노선 전체 길이와 같은 경우"),
            Arguments.of(101, "추가 구간의 길이가 노선 전체 길이보다 큰 경우")
        );
    }
```

parkeeseul replied:
@ParameterizedTest(name = "{index}: {1}")
각각 인자에 대한 테스트 명을 나타내 볼 수도 있을 것 같아요 😄

parkeeseul commented on src/main/java/nextstep/subway/dto/line/section/CreateSectionRequest.java:
```diff
@@ -0,0 +1,55 @@
+package nextstep.subway.dto.line.section;
+
+import java.util.ArrayList;
+import java.util.Arrays;
```
불필요한 import 는 제거해 주세요 😄

re-mora replied:
한번 더 확인했어야 했는데 😭 
해당 Import 구문은 제거하겠습니다.
꼼꼼하게 봐주셔서 감사합니다. 😄

parkeeseul commented on src/main/java/nextstep/subway/dto/line/section/CreateSectionRequest.java:
```diff
@@ -0,0 +1,55 @@
+            .filter(it -> it.getId().equals(downStationId))
+            .findFirst()
+            .orElseThrow(() -> new IllegalArgumentException("요청에 해당하는 하행역을 찾을 수 없습니다."));
+    }
```
DTO 는 구간 사이의  데이터 전송을 담당하는 객체입니다.
DTO는 기능을 가지지 않고 일종의 구조체로 사용되는데요
해당 로직은 service 에 위치하도록 해보면 어떨까요? 😄

re-mora replied:
service 계층에서 상태값을 가지는 객체의 getter를 사용하지 않으려 구현하다보니 DTO 객체에 로직을 작성하게 되었네요. 😭 
말씀하신 것처럼 DTO는 기능을 가지지 않도록 구현한 로직을 Service 계층으로 이동하였습니다 👍

parkeeseul commented on src/test/java/nextstep/subway/acceptance/SectionAcceptanceTest.java:
```diff
@@ -0,0 +1,201 @@
+
+@DisplayName("구간 관련 기능")
+@AcceptanceTest
+public class SectionAcceptanceTest {
```
BaseTest 를 상속받지 않은 이유가 있으신가요? 😄

re-mora replied:
자연스럽게.. 누락된 것 같습니다.

누락된 BaseTest 상속을 추가하였습니다!

parkeeseul commented on src/test/java/nextstep/subway/domain/line/LineTest.java:
```diff
@@ -0,0 +1,260 @@
+import org.junit.jupiter.params.provider.MethodSource;
+
+@DisplayName("Domain:Line")
+class LineTest {
```
테스트를 꼼꼼하게 작성해 주셨네요! 👍

parkeeseul commented on src/test/java/nextstep/subway/domain/line/SectionTest.java:
```diff
@@ -0,0 +1,91 @@
+
+    @Test
+    @DisplayName("상행역과 하행역이 동일한 구간 생성 시 예외")
+    public void ddd() {
```
@_@ ddd!! 😄

re-mora replied:
😭

parkeeseul commented on src/test/java/nextstep/subway/domain/line/SectionTest.java:
```diff
@@ -0,0 +1,91 @@
+    public void createSection() {
+        // Given
+        Station 논현역 = new Station("논현역");
+        Station 신논현역 = new Station("신논현역");
```
테스트에 자주 사용되는 객체들은
테스트 fixture 로 만들어 관리해 줘도 좋을 것 같아요 😄

parkeeseul commented on src/main/java/nextstep/subway/domain/line/wrap/Distance.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    private void validateDistance(int distance) {
+        if (distance <= 0) {
```
매직넘버를 사용하기 보다는 상수화를 통해 값에 의미를 부여해 보면 어떨까요? 😄

parkeeseul reviewed (changes requested):
3단계 미션 잘 진행해 주셨네요! 👍 
몇 가지 코멘트 남겨두었으니 확인 부탁드립니다 😄 
날씨가 많이 더워졌네요! 건강 유의하시고 미션 화이팅 입니다 🔥

re-mora committed:
docs: 코드 리뷰 피드백 내용 정리

re-mora committed:
fix: 잘못 작성된 구간 추가 API의 경로 수정

- AS-IS : /lines/{id}
- TO-BE : /lines/{id}/sections

re-mora committed:
chore(expression-improvement): 테스트 코드 표현 개선

- 구간 추가 인수테스트에 누락된 BaseTest 상속 추가
- `@DisplayName`에만 테스트를 표현하고 메소드에는 표현하지 않은 테스트명 개선
- 너무 많은 정보를 표현하다보니 오히려 가독성이 떨어진 테스트 표현 개선
- TC 작성을 위해 반복적으로 사용되는 객체 생성 부를 별도의 Fixture로 분리

re-mora committed:
chore(expression-improvement): 운영 코드 표현 개선

- 계층간 데이터 전송을 담당하는 DTO에 구현된 로직을 service 계층으로 이관
- 코드내 직접 명시된 매직넘버의 상수 처리
- 기타 불필요 코드 및 미사용 코드 제거

re-mora commented:
안녕하세요! :) 

코드 리뷰 해주신 아래 내용을 수정하여 한번 더 리뷰 요청 드립니다 :)

---

### 코드 리뷰 피드백 내용
잘못 작성된 구간 추가 API 경로 수정

테스트 코드 표현 개선
  - 구간 추가 인수테스트에 누락된 BaseTest 상속 추가
  - `@DisplayName`에만 테스트를 표현하고 메소드에는 표현하지 않은 테스트명 개선
  - 너무 많은 정보를 표현하다보니 오히려 가독성이 떨어진 테스트 표현 개선
  - TC 작성을 위해 반복적으로 사용되는 객체 생성 부를 별도의 Fixture로 분리

운영 코드 표현 개선
  - 계층간 데이터 전송을 담당하는 DTO에 구현된 로직을 service 계층으로 이관  
  - 코드내 직접 명시된 매직넘버의 상수 처리
  - 기타 불필요 코드 및 미사용 코드 제거

parkeeseul commented on src/main/java/nextstep/subway/domain/line/wrap/Distance.java:
```diff
@@ -8,6 +8,7 @@
 @Access(AccessType.FIELD)
 public class Distance {
 
+    public static final int MINIMUM_DISTANCE = 0;
```
`Distance.MINIMUM_DISTANCE` vs `Distance.MIN`
어떻게 생각하시나요 ? 😄

parkeeseul reviewed (approved):
피드백 잘 반영해 주셨네요! 👍 
소소한 코멘트 남겨두었으니 확인 부탁드리고
다음 단계 진행해 주세요! 😄 
화이팅 입니다 🔥
